commercial/app/model/merchandise/books/BookFinder.scala

```diff
@@ -2,67 +2,64 @@ package commercial.model.merchandise.books
-import shade.memcached.{Memcached, MemcachedCodecs, Configuration => MemcachedConfiguration}
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.control.NonFatal
 
 class BookFinder(actorSystem: ActorSystem, magentoService: MagentoService) extends Logging {
 
-  private implicit lazy val executionContext = actorSystem.dispatchers.lookup("akka.actor.memcached")
-
-  private lazy val defaultBookDataCache = new MemcachedBookDataCache(actorSystem)
-
-  def findByIsbn(isbn: String,
-                 cache: BookDataCache = defaultBookDataCache,
-                 lookup: String => Future[Option[JsValue]] = magentoService.findByIsbn):
-  Future[Option[Book]] = {
-
-    def cachedBook(bookData: JsValue): Future[Option[Book]] = {
-      Future.successful {
-        bookData match {
-          case JsNull => None
-          case _ =>
-            bookData.validate[Book] fold(
-              invalid => {
-                log.error(Json.stringify(JsError.toJson(invalid)))
-                None
-              },
-              book => Some(book))
-        }
+  private implicit val bookActorExecutionContext: ExecutionContext = actorSystem.dispatchers.lookup("akka.actor.book-lookup")
+  private implicit val bookActorTimeout: Timeout = 0.2.seconds
+  private implicit val magentoServiceImplicit = magentoService
+
+  def findByIsbn(isbn: String): Option[Book] = BookAgent.get(isbn) map { _.as[Book] }
+}
+
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val isbnsToBeAddedToCache = AkkaAgent(Set.empty[String])
+
+  def get(isbn: String)(implicit magentoService: MagentoService, executionContext: ExecutionContext): Option[JsValue] =
+
+    cache.get.get(isbn) match {
+      case Some(json) => {
+        log.info(s"Cache hit for ISBN: $isbn.")
+        Some(json)
       }
-    }
+      case None => {
+        log.debug(s"Cache miss for ISBN: $isbn.")
 
-    lazy val lookedUpBook = {
-      lookup(isbn) map {
-        _ map { bookData =>
-          cache.add(isbn, bookData)
-          bookData.as[Book]
-        } orElse {
-          cache.add(isbn, JsNull)
-          None
+        if (isbnsToBeAddedToCache.get.contains(isbn)) {
+          log.debug(s"ISBN lookup is already due for $isbn. Not running Magento Service.")
         }
-      }
-    }
+        else {
+          log.debug(s"Looking up ISBN $isbn with Magento Service.")
+          isbnsToBeAddedToCache alter { _ + isbn }
+
+          val lookup: Future[Option[JsValue]] = magentoService.findByIsbn(isbn)
 
-    cache.get(isbn) flatMap {
-      _ map cachedBook getOrElse lookedUpBook
-    } recoverWith {
-      case NonFatal(e) => lookedUpBook
+          lookup map {
+            case Some(json: JsValue) =>
+              cache alter { _ + (isbn -> json) }
+            case None =>
+              isbnsToBeAddedToCache alter { _ - isbn }
+          }
+        }
+        None
+      }
```

Comment:
Yep I can do that - though [Akka guarantees](http://doc.akka.io/docs/akka/current/scala/agents.html#Updating_Agents__send___alter_) that the order of operations of subsequent calls to `alter` on the same agent will be honured. So I don't think that it is strictly required here.